### PR TITLE
Add canMakePayments

### PR DESCRIPTION
**Pull request by beylmk**
Adds canMakePayments method

Opening in draft because I still need to remove the aars and clean up some formatting, but ready for review on the actual implementation.

### CODE REVIEW

File: RevenueCat/Scripts/Purchases.cs [hunk 1]
```diff
@@ -18,7 +18,8 @@ public partial class Purchases : MonoBehaviour
     public delegate void GetOfferingsFunc(Offerings offerings, Error error);
 
     public delegate void CheckTrialOrIntroductoryPriceEligibilityFunc(Dictionary<string, IntroEligibility> products);
-
+    
+    public delegate void CanMakePaymentsFunc(bool canMakePayments, Error error);
```
Comment (vegaro) [inline RevenueCat/Scripts/Purchases.cs:22]: I know we don't have docs for the other functions, but I think we should start adding them (and add for all the other functions on a Friday). Do you mind adding docs here?

File: RevenueCat/Scripts/PurchasesWrapperiOS.cs [hunk 2]
```diff
@@ -17,8 +17,7 @@ private class ProductsRequest
     {
         public string[] productIdentifiers;
     }
-
-    [DllImport("__Internal")]
+ [DllImport("__Internal")]
```
Comment (vegaro) [inline RevenueCat/Scripts/PurchasesWrapperiOS.cs:20]: I think this change is unintended

File: RevenueCat/Scripts/PurchasesWrapperiOS.cs [hunk 3]
```diff
@@ -29,6 +28,7 @@ public void GetProducts(string[] productIdentifiers, string type = "subs")
 
         _RCGetProducts(JsonUtility.ToJson(request), type);
     }
+
```
Comment (vegaro) [inline RevenueCat/Scripts/PurchasesWrapperiOS.cs:31]: extra line

**Review by vegaro — CHANGES_REQUESTED** [review]
looks good so far, I will re review when the tmp files are cleaned up

File: RevenueCat/Scripts/Purchases.cs (hunk 1 above)
Comment (beylmk) [reply RevenueCat/Scripts/Purchases.cs:22]: @vegaro created clubhouse ticket for the others: https://app.clubhouse.io/revenuecat/story/7152/add-documentation-to-unity-functions, as well as a ticket for researching/setting up unit testing (if you could add comments on that tool you saw, that'd be great...): https://app.clubhouse.io/revenuecat/story/7153/research-set-up-unity-unit-tests

File: Subtester/Assets/Plugins/Android/mainTemplate.gradle.DISABLED [hunk 4]
```diff
@@ -0,0 +1,72 @@
+// GENERATED BY UNITY. REMOVE THIS COMMENT TO PREVENT OVERWRITING WHEN EXPORTING AGAIN
```
Comment (aboedo) [inline Subtester/Assets/Plugins/Android/mainTemplate.gradle.DISABLED:1]: was checking in this file intended?

**Comment by beylmk** [comment]
@vegaro @aboedo i believe this is ready for re-review

File: RevenueCat/Plugins/Android/PurchasesWrapper.java [hunk 5]
```diff
@@ -320,6 +322,38 @@ public static void setCreative(String creative) {
...
+            }
+
+            CommonKt.canMakePayments(UnityPlayer.currentActivity, featuresToSend, new OnResultAny<Boolean>() {
+               @Override
+               public void onReceived(Boolean canMakePayments) {
+                   JSONObject object = new JSONObject();
+                   try {
+                       object.put("canMakePayments", canMakePayments);
+                   } catch (JSONException e) {
+                       e.printStackTrace();
```
Comment (aboedo) [inline RevenueCat/Plugins/Android/PurchasesWrapper.java:343]: why the difference in error handling between here and line 324?

**Review by aboedo — APPROVED** [review]

File: RevenueCat/Plugins/Android/PurchasesWrapper.java (hunk 5 above)
Comment (beylmk) [reply RevenueCat/Plugins/Android/PurchasesWrapper.java:343]: ah, no real reasoning just copy pasta. i'll update so they're consistent 👍

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m [hunk 6]
```diff
@@ -557,3 +567,21 @@ void _RCSetCreative(const char *creative) {
...
 }
+
+void _RCCanMakePayments(const char *featuresJSON) {
+    NSError *error = nil;
+    
+    NSDictionary *canMakePaymentsRequest = [NSJSONSerialization 
+        JSONObjectWithData:[convertCString(featuresJSON)        
+            dataUsingEncoding:NSUTF8StringEncoding]
+        options:0
+        error:&error];
```
Comment (vegaro) [inline RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:578]: The formatting of these lines is a little bit off

File: RevenueCat/Scripts/Purchases.cs [hunk 7]
```diff
@@ -403,6 +413,34 @@ public void CollectDeviceIdentifiers()
...
+     /// <summary>
+     /// Check if billing is supported for the current user (meaning IN-APP purchases are supported)
+     /// and whether a list of specified feature types are supported.
+     ///
+     /// Note: BillingFeatures are only relevant to Google Play Android users.
+     /// For other stores and platforms, BillingFeatures won't be checked.
+     /// </summary>
+     /// <param name="features">An array of BillingFeatures to check for support.
+     /// If empty, no features will be checked.</param>
+     /// <param name="callback">A callback receiving a bool for canMakePayments and potentially an Error</param>
```
Comment (vegaro) [inline RevenueCat/Scripts/Purchases.cs:428]: Great, love these docs thanks!!!

**Review by vegaro — APPROVED** [review]
Looks good! There's a Subtester/Assets/Plugins/Android/mainTemplate.gradle.DISABLED file that I am not sure is intended to be there :)

File: RevenueCat/Plugins/iOS/PurchasesUnityHelper.m (hunk 6 above)
Comment (beylmk) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:578]: there seemed to be a really wide array of formattings for this kinda call... this felt the most readable to me, but open to suggestions.

Comment (vegaro) [reply RevenueCat/Plugins/iOS/PurchasesUnityHelper.m:578]: In objc, `:` are aligned if the parameters are in different lines. Apart from that, maybe extracting the data into a variable also helps. Something like this :

```
NSData *data = [convertCString(featuresJSON) dataUsingEncoding:NSUTF8StringEncoding];
NSDictionary *canMakePaymentsRequest = [NSJSONSerialization JSONObjectWithData:data
                                                                       options:0
                                                                         error:&error];
```

**Pull request merged**